dm/pkg/router/router_test.go

```diff
@@ -0,0 +1,387 @@
+			Name:   fmt.Sprintf("dtable%d", i),
+		})
+	}
+	oldRouter, err := oldrouter.NewTableRouter(true, rules)
+	c.Assert(err, Equals, nil)
+	newRouter, err := NewRouter(true, rules)
+	c.Assert(err, Equals, nil)
+	for i := range inputTables {
+		schema, table := inputTables[i].Schema, inputTables[i].Name
+		expSchema, expTable := expTables[i].Schema, expTables[i].Name
+		oldSch, oldTbl, _ := oldRouter.Route(schema, table)
+		newSch, newTbl, _ := newRouter.Route(schema, table)
+		c.Assert(newSch, Equals, expSchema)
+		c.Assert(newTbl, Equals, expTable)
+		c.Assert(oldSch, Equals, expSchema)
+		c.Assert(oldTbl, Equals, expTable)
+	}
+}
+
+func (s *testRouteSuite) TestRegExprRoute(c *C) {
+	rules := []*TableRule{
+		{
+			SchemaPattern: "~test.[0-9]+",
+			TargetSchema:  "dtest1",
+		},
+		{
+			SchemaPattern: "~test2?[ab|cd]",
+			TablePattern:  "~g*[cat|dog]+[0-9]*",
+			TargetSchema:  "dtest2",
+			TargetTable:   "dtable2",
+		},
+		{
+			SchemaPattern: "~test3(gd|ct)?8917*",
+			TablePattern:  "test3_*",
+			TargetSchema:  "dtest3",
+			TargetTable:   "dtable3",
+		},
+		{
+			SchemaPattern: "test4s_*",
+			TablePattern:  "~testtable4?(ddd|lbc)+",
+			TargetSchema:  "dtest4",
+			TargetTable:   "dtable4",
+		},
+	}
+	inputTable := []Table{
+		{
+			Schema: "tests99",
+			Name:   "table1", // match rule 1
+		},
+		{
+			Schema: "testcd",
+			Name:   "ggggdog9897", // match rule 2
+		},
+		{
+			Schema: "test3gd8917777",
+			Name:   "test3_99aa99", // match rule 3
+		},
+		{
+			Schema: "test4s_2022",
+			Name:   "testtablelbclbc", // match rule 4
+		},
+		{
+			Schema: "mytst5566",
+			Name:   "gtable", // match nothing
+		},
+	}
+	expectedOutput := []Table{
+		{
+			Schema: "dtest1",
+			Name:   "table1",
+		},
+		{
+			Schema: "dtest2",
+			Name:   "dtable2",
+		},
+		{
+			Schema: "dtest3",
+			Name:   "dtable3",
+		},
+		{
+			Schema: "dtest4",
+			Name:   "dtable4",
+		},
+		{
+			Schema: "mytst5566",
+			Name:   "gtable",
+		},
+	}
+	newRouter, err := NewRouter(true, rules)
+	c.Assert(err, Equals, nil)
+	for idx := range inputTable {
+		s, n := inputTable[idx].Schema, inputTable[idx].Name
+		expSchm, expName := expectedOutput[idx].Schema, expectedOutput[idx].Name
+		newSchm, newName, err := newRouter.Route(s, n)
+		c.Assert(err, Equals, nil)
+		c.Assert(newSchm, Equals, expSchm)
+		c.Assert(newName, Equals, expName)
+	}
+}
+
+func (s *testRouteSuite) TestFetchExtendColumn(c *C) {
+	rules := []*TableRule{
+		{
+			SchemaPattern: "schema*",
+			TablePattern:  "t*",
+			TargetSchema:  "test",
+			TargetTable:   "t",
+			TableExtractor: &TableExtractor{
+				TargetColumn: "table_name",
+				TableRegexp:  "table_(.*)",
+			},
+			SchemaExtractor: &SchemaExtractor{
+				TargetColumn: "schema_name",
+				SchemaRegexp: "schema_(.*)",
+			},
+			SourceExtractor: &SourceExtractor{
+				TargetColumn: "source_name",
+				SourceRegexp: "source_(.*)_(.*)",
+			},
+		},
+		{
+			SchemaPattern: "~s?chema.*",
+			TargetSchema:  "test",
+			TargetTable:   "t2",
```

Comment:
this usage (no table-pattern but target-table) is strange, should we forbid it in future? @GMHDBJD @lichunzhu

Reply:
IMO we should forbid this kind of behavior

Reply:
What if the users would like to migrate all tables from one or more designated schema(s) to exactly **ONE** table in downstream?

Reply:
They should use TablePattern: "*"

Reply:
But for now, missing field `TablePattern` is permitted. Would it be troublesome if we forbid it, considering some users may need to modify their old configs?